codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/XmlShapeDeserVisitor.java

```diff
@@ -0,0 +1,403 @@
+import software.amazon.smithy.model.traits.XmlNameTrait;
+import software.amazon.smithy.utils.FunctionalUtils;
+
+/**
+ * Visitor to generate deserialization functions for shapes in XML protocol
+ * document bodies.
+ *
+ * This class handles function body generation for all types expected by the
+ * {@code DocumentShapeDeserVisitor}. No other shape type serialization is overwritten.
+ *
+ * Timestamps are serialized to {@link TimestampFormatTrait.Format}.DATE_TIME by default.
+ */
+public class XmlShapeDeserVisitor extends DocumentShapeDeserVisitor {
+    private static final TimestampFormatTrait.Format DEFAULT_TIMESTAMP_FORMAT = TimestampFormatTrait.Format.DATE_TIME;
+    private static final Logger LOGGER = Logger.getLogger(XmlShapeDeserVisitor.class.getName());
+
+    private final Predicate<MemberShape> memberFilter;
+
+    /**
+     * @param context The generation context.
+     */
+    public XmlShapeDeserVisitor(GenerationContext context) {
+        this(context, FunctionalUtils.alwaysTrue());
+    }
+
+    /**
+     * @param context The generation context.
+     * @param memberFilter A filter that is applied to structure members. This is useful for
+     *     members that won't be in the body.
+     */
+    public XmlShapeDeserVisitor(GenerationContext context, Predicate<MemberShape> memberFilter) {
+        super(context);
+        this.memberFilter = memberFilter;
+    }
+
+    @Override
+    protected Map<String, String> getAdditionalArguments() {
+        return Collections.singletonMap("decoder", "*smithydecoding.XMLNodeDecoder");
+    }
+
+    private XmlMemberDeserVisitor getMemberDeserVisitor(MemberShape member, String dataDest, Predicate<Shape> filter) {
+        // Get the timestamp format to be used, defaulting to epoch seconds.
+        TimestampFormatTrait.Format format = member.getMemberTrait(getContext().getModel(), TimestampFormatTrait.class)
+                .map(TimestampFormatTrait::getFormat).orElse(DEFAULT_TIMESTAMP_FORMAT);
+
+        // Check if FlattenedTrait  is applied
+        Boolean isFlattened = member.hasTrait(XmlFlattenedTrait.ID);
+
+        return new XmlMemberDeserVisitor(getContext(), dataDest, format, filter, isFlattened);
+    }
+
+    @Override
+    protected void deserializeCollection(GenerationContext context, CollectionShape shape) {
+        GoWriter writer = context.getWriter();
+        SymbolProvider symbolProvider = context.getSymbolProvider();
+        Symbol symbol = symbolProvider.toSymbol(shape);
+
+        // Initialize the value now that the start stub has verified that there's something there.
+        writer.write("var sv $P", symbol);
+        writer.openBlock("if *v == nil {", "", () -> {
+            writer.write("sv = make($P, 0)", symbol);
+            writer.openBlock("} else {", "}", () -> {
+                writer.write("sv = *v");
+            });
+        });
+
+        // Iterate through the decoder. The member visitor will handle popping tokens.
```

Comment:
This does not seem to be entirely true as you are popping a token on each iteration, unless `.Token` doesn't consume in XML